frontend/projects/upgrade/src/app/core/stratification-factors/store/stratification-factors.effects.ts

```diff
@@ -0,0 +1,105 @@
+import { Injectable } from '@angular/core';
+import { Router } from '@angular/router';
+import { Actions, createEffect, ofType } from '@ngrx/effects';
+import { map, switchMap, filter, catchError, withLatestFrom } from 'rxjs/operators';
+import { Store, select } from '@ngrx/store';
+import { AppState } from '../../core.state';
+import * as StratificationFactorsActions from './stratification-factors.actions';
+import { StratificationFactorsDataService } from '../stratification-factors.data.service';
+import { selectAllStratificationFactors } from './stratification-factors.selectors';
+
+@Injectable()
+export class StratificationFactorsEffects {
+  constructor(
+    private actions$: Actions,
+    private router: Router,
+    private store$: Store<AppState>,
+    private stratificationFactorsDataService: StratificationFactorsDataService
+  ) {}
+
+  fetchStratificationFactors$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(StratificationFactorsActions.actionFetchStratificationFactors),
+      withLatestFrom(this.store$.pipe(select(selectAllStratificationFactors))),
+      switchMap(() =>
+        this.stratificationFactorsDataService.fetchStratificationFactors().pipe(
+          map((data: any) =>
+            StratificationFactorsActions.actionFetchStratificationFactorsSuccess({
+              stratificationFactors: data,
+            })
+          ),
+          catchError(() => [StratificationFactorsActions.actionFetchStratificationFactorsFailure()])
+        )
+      )
+    )
+  );
+
+  deleteStratificationFactor$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(StratificationFactorsActions.actionDeleteStratificationFactor),
+      map((action) => action.factor),
+      filter((factor) => !!factor),
+      switchMap((factor) =>
+        this.stratificationFactorsDataService.deleteStratificationFactor(factor).pipe(
+          map((data: any) => {
+            this.router.navigate(['/participants']);
```

Comment:
Not sure how we've done it elsewhere, but just eyeballing this it feel like we're navigating away before we're really finished executing code; I think this will may work out ok, but I think it would probably be cleanest to dispatch a separate action for navigating once everything else has finished executing. `actionDeleteStratificationFactorSuccess` seems like a more natural place to do navigation.

Reply:
We have been navigating in effects at all other places as well. If we want to move it to actions, we will have to do it for all.

Reply:
navigating in the effects is fine, I'm just asking for that to happen last, when we're executing "success" actions; it's confusing to navigate away before we're finished executing other code.

Reply:
Okay, got it

Reply:
@danoswaltCL So, I would update the other component's navigations in a separate PR. Hope this is what you want:

<img width="798" alt="Screenshot 2023-10-19 at 12 30 23 PM" src="https://github.com/CarnegieLearningWeb/UpGrade/assets/33730817/46baa667-7c48-4ff8-bfc2-18898f1982ad">

Reply:
@danoswaltCL Are we good here?